### ISSUE TITLE
libbpf: Switch kernel version lookup to a more reliable method.

### ISSUE THREAD

**Issue by usamasaqib** [libbpf/libbpf#421]
On older kernel versions, the kernel checks whether the version number provided in the `bpf_attr` object, matches the LINUX_VERSION_CODE (https://elixir.bootlin.com/linux/v4.19/source/kernel/bpf/syscall.c#L1360). If not, an error code is returned and the loading fails.
I have an encountered issues where the kernel version number reported by the uname syscall does not match the LINUX_VERSION_CODE.
Following is the example output of uname syscall on system where the LINUX_VERSION_CODE is 4.19.37
> sysname: Linux
> nodename: pardus
> release: 4.19.0-5-amd64
> version: #1 SMP Debian 4.19.37-5+deb10u1 (2019-07-19)
> machine: x86_64

Since libbpf uses the uname syscall to probe the kernel version, it fails to load BPF programs on such environments.

Bpftrace attempts to solve this issue by parsing the kernel version code from the PT_NOTE section of the vdso file mapped into the virtual memory of every program. This seems to contain the correct kernel version number matching LINUX_VERSION_CODE.
https://github.com/iovisor/bpftrace/blob/master/src/utils.cpp#L886
For architectures where this mapping is not present, it defaults to the uname syscall.

Maybe a similar approach can be replicated in libbpf.

**Comment by anakryiko** [comment]
perf seems to be using a different approach, it's parsing /proc/version_signature, see https://github.com/torvalds/linux/blob/master/tools/perf/util/util.c#L347

I wonder which one is more reliable. I don't have an Ubuntu system to test something like this, can you please post the contents of `/proc/version_signature` here?

**Comment by usamasaqib** [comment]
Output of uname -a is:

> Linux usama 5.4.0-91-generic #102~18.04.1-Ubuntu SMP Thu Nov 11 14:46:36 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Output of `cat /proc/version_signature` is:

> Ubuntu 5.4.0-91.102~18.04.1-generic 5.4.151

Kernel version obtained from parsing the vdso section:

> release: 5.4.151

**Comment by Jongy** [comment]
FWIW I've hit a kernel where the current method fails.

```
root@(none):/app# cat /proc/version 
Linux version 4.19.0-21-amd64 (debian-kernel@lists.debian.org) (gcc version 8.3.0 (Debian 8.3.0-6)) #1 SMP Debian 4.19.249-2 (2022-06-30)
root@(none):/app# uname -a
Linux (none) 4.19.0-21-amd64 #1 SMP Debian 4.19.249-2 (2022-06-30) x86_64 GNU/Linux
```

(downloaded [here](https://debian.sipwise.com/debian-security/pool/main/l/linux/linux-image-4.19.0-21-amd64-unsigned_4.19.249-2_amd64.deb); vmlinuz sha1 `021e526222a9bdd85c44e42d08bdb1c710a8b1d7`)

libbpf extracts 4.19.249 while the kernel expects 4.19.255 (I found it by disassembling `bpf_prog_load`).

The vdso parsing seems to work, however. I tried the [logic from bpftrace](https://github.com/bpftrace/bpftrace/blob/b530dd8accc91cba4d1886fd5d65f5d4bf36219f/src/util/kernel.cpp#L33) and it extracts 4.19.255 correctly, with which my kprobes load fine.

This is all very legacy, but @anakryiko if you consider it worthy I can send a patch to lkml that prefers the vdso method when it works. Or maybe only for Debian.
